src/smbclient/_os.py

```diff
@@ -306,17 +306,208 @@ def makedirs(path, exist_ok=False, **kwargs):
+FileType: t.TypeAlias = t.Literal["file", "dir", "pipe"]
+
+
+class SMBBufferedRandom(io.BufferedRandom):
+    raw: SMBRawIO  # type: ignore
+
+
+class SMBBufferedReader(io.BufferedReader):
+    raw: SMBRawIO  # type: ignore
+
+
+class SMBBufferedWriter(io.BufferedWriter):
+    raw: SMBRawIO  # type: ignore
+
+
+# Text mode: always returns a TextIOWrapper
+@t.overload
+def open_file(
+    path,
+    mode: OpenTextMode = "r",
+    buffering=-1,
+    file_type: FileType = "file",
+    encoding=None,
+    errors=None,
+    newline=None,
+    share_access=None,
+    desired_access=None,
+    file_attributes=None,
+    **kwargs,
+) -> io.TextIOWrapper[SMBBufferedRandom | SMBBufferedReader | SMBBufferedWriter]: ...
+
+
+# Unbuffered binary mode: returns a File/Directory/Pipe IO object
+@t.overload
+def open_file(
+    path,
+    mode: OpenBinaryMode,
+    buffering: t.Literal[0],
+    file_type: t.Literal["file"] = "file",
+    encoding=None,
+    errors=None,
+    newline=None,
+    share_access=None,
+    desired_access=None,
+    file_attributes=None,
+    **kwargs,
+) -> SMBFileIO: ...
+@t.overload
+def open_file(
+    path,
+    mode: OpenBinaryMode,
+    buffering: t.Literal[0],
+    file_type: t.Literal["dir"],
+    encoding=None,
+    errors=None,
+    newline=None,
+    share_access=None,
+    desired_access=None,
+    file_attributes=None,
+    **kwargs,
+) -> SMBDirectoryIO: ...
+@t.overload
+def open_file(
+    path,
+    mode: OpenBinaryMode,
+    buffering: t.Literal[0],
+    file_type: t.Literal["pipe"],
+    encoding=None,
+    errors=None,
+    newline=None,
+    share_access=None,
+    desired_access=None,
+    file_attributes=None,
+    **kwargs,
+) -> SMBPipeIO: ...
+
+
+# Buffering is on: return BufferedRandom, BufferedReader, or BufferedWriter
+# NOTE: This cannot handle explicit buffer sizes (>0) because of limitations
+# in Python's type system
+@t.overload
+def open_file(
+    path,
+    mode: OpenBinaryModeUpdating,
+    buffering: t.Literal[-1] = -1,
+    file_type: FileType = "file",
+    encoding=None,
+    errors=None,
+    newline=None,
+    share_access=None,
+    desired_access=None,
+    file_attributes=None,
+    **kwargs,
+) -> SMBBufferedRandom: ...
+@t.overload
+def open_file(
+    path,
+    mode: OpenBinaryModeReading,
+    buffering: t.Literal[-1] = -1,
+    file_type: FileType = "file",
+    encoding=None,
+    errors=None,
+    newline=None,
+    share_access=None,
+    desired_access=None,
+    file_attributes=None,
+    **kwargs,
+) -> SMBBufferedReader: ...
+@t.overload
+def open_file(
+    path,
+    mode: OpenBinaryModeWriting,
+    buffering: t.Literal[-1] = -1,
+    file_type: FileType = "file",
+    encoding=None,
+    errors=None,
+    newline=None,
+    share_access=None,
+    desired_access=None,
+    file_attributes=None,
+    **kwargs,
+) -> SMBBufferedWriter: ...
+
+
 def open_file(
     path,
     mode="r",
     buffering=-1,
+    file_type: t.Literal["file", "dir", "pipe"] = "file",
```

Comment:
So why does this need to be defined before the others? Is it because the overload only works based on the argument order?

Reply:
I think let's just ignore the unbuffered file types and we can investigate a better API in the future without hopefully exposing a breaking change.